ethereumj-core/src/main/java/org/ethereum/listener/RecommendedGasPriceTracker.java

```diff
@@ -0,0 +1,179 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The ethereumJ library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with the ethereumJ library. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.ethereum.listener;
+
+import org.apache.commons.collections4.queue.CircularFifoQueue;
+import org.ethereum.core.Block;
+import org.ethereum.core.BlockSummary;
+import org.ethereum.core.Transaction;
+import org.ethereum.util.ByteUtil;
+
+import java.lang.reflect.Array;
+import java.util.Arrays;
+import java.util.List;
+
+
+/**
+ * Calculates a 'reasonable' Gas price based on statistics of the latest transaction's Gas prices.
+ * This is an updated version that returns more accurate data
+ * in networks with large number of transactions like Ethereum MainNet.
+ * However it needs more CPU and memory resources for processing.
+ *
+ * Normally the price returned should be sufficient to execute a transaction since ~25%
+ * (if {@link #getPercentileShare()} is not overridden) of the latest transactions were
+ * executed at this or lower price.
+ */
+public class RecommendedGasPriceTracker extends EthereumListenerAdapter {
+
+    private static final Long DEFAULT_PRICE = null;
+    private static final int MIN_BLOCKS = 128;
+    private static final int BLOCKS_RECOUNT = 1;
+    private static final int MIN_TRANSACTIONS = 512;
+    private static final int PERCENTILE_SHARE = 4;
+
+    private CircularFifoQueue<long[]> blockGasPrices;
+
+    private int idx = 0;
+    private Long recommendedGasPrice = getDefaultPrice();
+
+    public RecommendedGasPriceTracker() {
+        blockGasPrices = new CircularFifoQueue<>(Math.max(getMinTransactions(), getMinBlocks()));
+    }
+
+    @Override
+    public void onBlock(BlockSummary blockSummary) {
+        onBlock(blockSummary.getBlock());
+    }
+
+    private void onBlock(Block block) {
+        if (onTransactions(block.getTransactionsList())) {
+            ++idx;
+            if (idx == getBlocksRecount()) {
+                Long newGasPrice = getGasPrice();
+                if (newGasPrice != null) {
+                    this.recommendedGasPrice = newGasPrice;
+                }
+                idx = 0;
+            }
+        }
+    }
+
+    private synchronized boolean onTransactions(List<Transaction> txs) {
+        if (txs.isEmpty()) return false;
+
+        long[] gasPrices = new long[txs.size()];
+        for (int i = 0; i < txs.size(); ++i) {
+            gasPrices[i] = ByteUtil.byteArrayToLong(txs.get(i).getGasPrice());
+        }
+
+        while (blockGasPrices.size() >= getMinBlocks() &&
```

Comment:
If `gasPrices` would be added before the loop then we got rid of extra `+ gasPrices.length` argument in the loop condition. But this would only work if intrusive `CircularFifoQueue` was replaced

Reply:
If above changes are applied `>=` can be replaced with  `>` which should be more precise condition